File: sample/src/main/java/uk/co/barbuzz/chompprogressview/sample/MainActivity.java
package uk.co.barbuzz.chompprogressview.sample;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import uk.co.barbuzz.chompprogressview.ChompProgressImageView;

public class MainActivity extends AppCompatActivity implements ChompNoshThread.ChompNoshListener {

    private static final String TAG = "MainActivity";

    private static final int BITE_SIZE_PIZZA = 440;
    private static final int BITE_SIZE_DONUT = 470;
    private static final int BITE_SIZE_ICE_CREAM = 290;

    private BottomNavigationView mBottomBar;
    private ChompProgressImageView mChompProgressImageView;
    private AlertDialog infoDialog;
    private Drawable mPizzaDrawable, mDonutDrawable, mIceCreamDrawable;
    private ChompNoshThread mChompNoshThread;
    private int mMenuId = R.id.pizza_item;

    private View.OnClickListener mImageClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            if (mMenuId == R.id.pizza_item) {
                eatNosh(BITE_SIZE_PIZZA, mPizzaDrawable, false);
            } else if (mMenuId == R.id.donut_item) {
                eatNosh(BITE_SIZE_DONUT, mDonutDrawable, false);
            } else if (mMenuId == R.id.ice_cream_item) {
                eatNosh(BITE_SIZE_ICE_CREAM, mIceCreamDrawable, true);
            }
            mChompProgressImageView.setOnClickListener(null);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_github) {
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse(getResources().getString(R.string.github_link)));
            startActivity(i);
            return true;
        } else if (id == R.id.action_info) {
            showInfoDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onChompNoshProgress(int progress) {
        mChompProgressImageView.setChompProgress(progress);
    }

    @Override
    public void onChompNoshFinished() {
        //reset chomp image view
        mChompProgressImageView.removeBites();
        mChompProgressImageView.setChompProgress(0);
        mChompProgressImageView.setTotalNumberOfBitesTaken(0);
        //re-enable image view onclick
        mChompProgressImageView.setOnClickListener(mImageClickListener);
    }

    private void showInfoDialog() {
        if (infoDialog != null && infoDialog.isShowing()) {
            //do nothing if already showing
        } else {
            infoDialog = new AlertDialog.Builder(this)
                    .setTitle(R.string.app_name)
                    .setMessage(R.string.info_details)
                    .setCancelable(true)
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                        }
                    })
                    .setNegativeButton("More info", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                            startActivity(new Intent(Intent.ACTION_VIEW,
                                    Uri.parse(getResources().getString(R.string.github_link))));
                        }
                    })
                    .create();
            infoDialog.show();
        }
    }

    private void initViews() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mPizzaDrawable = getResources().getDrawable(R.drawable.pizza);
        mDonutDrawable = getResources().getDrawable(R.drawable.donut);
        mIceCreamDrawable = getResources().getDrawable(R.drawable.icecream);

        mChompProgressImageView = findViewById(R.id.content_main_chomp_progress_imageview);
        mChompProgressImageView.setOnClickListener(mImageClickListener);

        mBottomBar = findViewById(R.id.bottom_navigation);
        mBottomBar.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                //cancel last chomp progress task and store menu id for onclick of image
                stopEating();
                //store menu id so we can use it in image onClick
                mMenuId = item.getItemId();
                switch (mMenuId) {
                    case R.id.pizza_item:
                        mChompProgressImageView.setImageDrawableChomp(mPizzaDrawable);
                        break;
                    case R.id.donut_item:
                        mChompProgressImageView.setImageDrawableChomp(mDonutDrawable);
                        break;
                    case R.id.ice_cream_item:
                        mChompProgressImageView.setImageDrawableChomp(mIceCreamDrawable);
                        break;
                }
                mChompProgressImageView.setOnClickListener(mImageClickListener);
                return true;
            }
        });
    }

    private void eatNosh(int biteSize, Drawable imageDrawable, boolean isChompFromTop) {
        mChompProgressImageView.setImageDrawableChomp(imageDrawable);
        mChompProgressImageView.setChompDirection(isChompFromTop ?
                ChompProgressImageView.ChompDirection.TOP :
                ChompProgressImageView.ChompDirection.RANDOM);
        mChompProgressImageView.setBiteRadius(biteSize);

        mChompNoshThread = new ChompNoshThread(this);
        mChompNoshThread.start();
    }

    private void stopEating() {
        if (mChompNoshThread != null) {
            mChompNoshThread.terminate();
        }
    }
}

File: sample/src/main/java/uk/co/barbuzz/chompprogressview/sample/MainActivity.kt
package uk.co.barbuzz.chompprogressview.sample

import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import uk.co.barbuzz.chompprogressview.ChompProgressImageView

class MainActivity : AppCompatActivity(), ChompNoshThread.ChompNoshListener {

    private lateinit var infoDialog: AlertDialog
    private lateinit var pizzaDrawable: Drawable
    private lateinit var donutDrawable: Drawable
    private lateinit var iceCreamDrawable: Drawable
    private var chompNoshThread: ChompNoshThread? = null
    private var menuId = R.id.pizza_item

    private val imageClickListener = View.OnClickListener {
        when (menuId) {
            R.id.pizza_item -> eatNosh(BITE_SIZE_PIZZA, pizzaDrawable, false)
            R.id.donut_item -> eatNosh(BITE_SIZE_DONUT, donutDrawable, false)
            R.id.ice_cream_item -> eatNosh(BITE_SIZE_ICE_CREAM, iceCreamDrawable, true)
        }
        content_main_chomp_progress_imageview.setOnClickListener(null)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        initViews()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_github) {
            val i = Intent(Intent.ACTION_VIEW)
            i.data = Uri.parse(resources.getString(R.string.github_link))
            startActivity(i)
            return true
        } else if (id == R.id.action_info) {
            showInfoDialog()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onChompNoshProgress(progress: Int) {
        content_main_chomp_progress_imageview.chompProgress = progress
    }

    override fun onChompNoshFinished() {
        //reset chomp image view
        content_main_chomp_progress_imageview.removeBites()
        content_main_chomp_progress_imageview.chompProgress = 0
        content_main_chomp_progress_imageview.totalNumberOfBitesTaken = 0
        //re-enable image view onclick
        content_main_chomp_progress_imageview.setOnClickListener(imageClickListener)
    }

    private fun showInfoDialog() {
        if (infoDialog.isShowing) {
            //do nothing if already showing
        } else {
            infoDialog = AlertDialog.Builder(this)
                    .setTitle(R.string.app_name)
                    .setMessage(R.string.info_details)
                    .setCancelable(true)
                    .setPositiveButton("OK") { dialog, which -> dialog.dismiss() }
                    .setNegativeButton("More info") { dialog, which ->
                        dialog.dismiss()
                        startActivity(Intent(Intent.ACTION_VIEW,
                                Uri.parse(resources.getString(R.string.github_link))))
                    }
                    .create()
            infoDialog.show()
        }
    }

    private fun initViews() {
        setSupportActionBar(toolbar)

        pizzaDrawable = resources.getDrawable(R.drawable.pizza)
        donutDrawable = resources.getDrawable(R.drawable.donut)
        iceCreamDrawable = resources.getDrawable(R.drawable.icecream)

        content_main_chomp_progress_imageview.setOnClickListener(imageClickListener)

        bottom_navigation.setOnNavigationItemSelectedListener { item ->
            //cancel last chomp progress task and store menu id for onclick of image
            stopEating()
            //store menu id so we can use it in image onClick
            menuId = item.itemId
            when (menuId) {
                R.id.pizza_item -> content_main_chomp_progress_imageview.setImageDrawableChomp(pizzaDrawable)
                R.id.donut_item -> content_main_chomp_progress_imageview.setImageDrawableChomp(donutDrawable)
                R.id.ice_cream_item -> content_main_chomp_progress_imageview.setImageDrawableChomp(iceCreamDrawable)
            }
            content_main_chomp_progress_imageview.setOnClickListener(imageClickListener)
            true
        }
    }

    private fun eatNosh(biteSize: Int, imageDrawable: Drawable?, isChompFromTop: Boolean) {
        content_main_chomp_progress_imageview.setImageDrawableChomp(imageDrawable)
        content_main_chomp_progress_imageview.setChompDirection(if (isChompFromTop)
            ChompProgressImageView.ChompDirection.TOP
        else
            ChompProgressImageView.ChompDirection.RANDOM)
        content_main_chomp_progress_imageview.biteRadius = biteSize

        chompNoshThread = ChompNoshThread(this)
        chompNoshThread?.start()
    }

    private fun stopEating() {
        chompNoshThread?.terminate()
    }

    companion object {
        private const val BITE_SIZE_PIZZA = 440
        private const val BITE_SIZE_DONUT = 470
        private const val BITE_SIZE_ICE_CREAM = 290
    }
}
